### ISSUE TITLE
Store variable with level script to access in later level

### ISSUE THREAD

**Issue by Loobinex** [Loobinex/keeperfx-unofficial#38]
As a mapmaker I want to have what the player does in one level affect how a future level plays out in the same campaign.
To accomplish this I want to be able to store a variable with the level script, and be able to compare that variable with pre-scripted outcomes in the other.

**Comment by OrsonMaxwell** [comment]
> There are two things that bullfrog did better with the SET_FLAG command:

> - The name of the variable (FLAG1) is used both when you set it AND when you compare it. This way anybody who reads the script can understand it without checking the manual.
> - The flag is linked to a specific player both on setting AND on comparing the flag. Without this, the scripter would never guess/understand why.

I actually made it work exactly like the flags (using the full names of the variables), but opted out of it because it bulks up the code. However this one can be reverted because mapmakers do not see the code - they see the scripts, so it may be more important to maintain consistency here.
Player assignment of variables does not make sense to me in the context of campaigns. It would make sense for flags because there are only 8 of them for each player. We can mimic this, but in my understanding we can introduce confusion while chasing consistency. If I see something like this:
```
SET_CAMPAIGN_VARIABLE(PLAYER0, CAMPAIGN_VARIABLE1, 42)
```
I immediately think two things:

- There is an implied connection between the player and the stored value, but how does it manifest itself? In other words, what does "PLAYER0" actually mean in the context of a campaign which is single-player by definition?
- "CAMPAIGN_VARIABLE" is spelled out twice in the same line which adds to neither brevity nor clarity.

So those were my motivations to have simpler syntax, albeit non-conforming to the existing FLAG syntax. Campaign variables are not flags - they have entirely different purpose and underlying mechanism and therefore I didn't try very hard to unify its syntax with the flags.
To be honest, the original DK flag idea is itself ambiguous: every person who tried programming on any language at all instinctively consider flags to be boolean variable. This brings me to the next point:

> This simply accepts 'FLAG' as the dungeon keeper name for 'variable', and at least avoids confusion on what we're talking about.
This actually adds to the confusion because we start to use one name (and not a really great one) for different entities.

Having said all that I strongly gravitate towards option number **2**.

Human interfacing API is, of course, important, and I agree that making it consistent with the rest of the scripting language is important. However, let us first define what we **cannot do** in the current state of the codebase:
- We cannot directly extend flag range - we can only hack in some code so it looks like it;
- We cannot override engine variables like MONEY, TOTAL_AREA etc. with our value;
- We cannot store values of these variables whenever we want to (AFAIK).

**Comment by Loobinex** [comment]
Yes a fourth possibility would be to make it like this by the way:
```
SET_FLAG(PLAYER0,FLAG1,10)
IF(CAMPAIGN,FLAG1 == 10)
       WIN_GAME
ENDIF
```

I also don't think that first option is the best. Yesterday I had a good look at all the commands you introduced, and because I had to try really hard to even understand the new introduced commands. And in the end it was unified, but I think it introduced too many problems. It was time for bed, so posted as far as I got, but I don't think we should go this route.
Just using flags would be such an elegant solution, but the having to manually load/save them for use in next levels just sucks, just as alsways saving/loading them without player choice and forcing mapmakers to clear each flag each level when they don't want to.

In any case, it seems like the additional commands you thought up, won't be introduced anyway, and the one you do have works already. So to keep it simple for ourselves, we best stick close to what we have already.
I have to say though, I'm kinda unhappy with the 'campaign_variable' name, it's just that I haven't be able to think of something better. Simply 'Variable' might be an option, or do stick with 'CAMVAR' like you proposed or a variant on 'flag'. What do you think about 'CAMPAIGN_FLAG', and perfectly duplicating the FLAG command in all other ways?
And that includes splitting the 32 variables over 4 players yes, is there any technical reason not to link it to a specific player?
Like regular flags, it does not really serve much purpose, since most things happen to player0 anyway, but I can easily assume that I set the PLAYER1,CAMPAIGN_FLAG1 to 1 if I defeated player1 on a certain campaign.

As for including the

**Comment by Loobinex** [comment]
Fixed with 1f50640b11e7e3631b02e17333820d38f5bee49f
